File: phoneBooks/src/phoneBooks/PhoneBookExt.java
package phoneBooks;

import bookIOFile.MatchForPhone;

public class PhoneBookExt extends PhoneBook{
    Telephone contactsPhones = new Telephone();

    public PhoneBookExt() {
        super();
    }

    public void addToBook(Contact contact) {
        getContacts().add(contact);
        for (String t: contact.getPhone().getTelephones()) {
            contactsPhones.addTelephone(t);
        }
    }

    @Override
    public void addContact(Contact contact) {
        if (contactsPhones.phoneCount() == 0 ||
                !getContacts().contains(contact)) {
            if (isHasPhone(contact.getPhone())) {
                for (String t : contact.getPhone().getTelephones()) {
                    Contact tmp = searchByTelephone(t);
                    if (tmp != null) {
                        MatchForPhone.writeError(tmp, contact,
                                String.format("Телефон %s уже есть!", t));
                        return;
                    }
                }
            }
            addToBook(contact);
        } else {
            Contact tmp = searchByName(contact.getFullName());
            if (tmp.getPhone().phoneCount() < contact
                    .getPhone().phoneCount()) {
                removeContact(tmp);
                addContact(contact);
                if (!getContacts().contains(contact)) {
                    addContact(tmp);
                }
            }
//            MatchForPhone.writeError(tmp, contact,
//                    "Контакт уже есть!");
        }
    }

    @Override
    public Contact searchByTelephone(String phone) {
        for (Contact c : getContacts()) {
            if (c.getPhone().isTelephone(phone)) {
                return c;
            }
        }
        return null;
    }

    public  boolean isHasPhone(Telephone tp) {
        for (String t : tp.getTelephones()) {
            if (contactsPhones.isTelephone(t)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean removeContact(Contact contact) {
        if (getContacts().contains(contact)) {
            for (String t : contact.getPhone().getTelephones()) {
                contactsPhones.removePhone(t);
            }
            super.removeContact(contact);
            return true;
        }
        return false;
    }
}

File: phoneBook/phoneBook/PhoneBookExt.cs
using phoneBook.bookIOFile;

namespace phoneBook.phoneBook
{
    public class PhoneBookExt : PhoneBook
    {
        private Telephone contactsPhones = new Telephone();

        public bool IsPhone(Contact contact)
        {
            foreach (string t in contact.Phone.Telephones)
            {
                if (contactsPhones.IsPhone(t))
                {
                    return true;
                }
            }
            return false;
        }

        public string GetPhone(Contact contact)
        {
            string str = string.Empty;
            foreach (string item in contact.Phone.Telephones)
            {
                if (contactsPhones.IsPhone(item))
                {
                    str = item;
                }
            }
            return str;
        }

        public override bool Add(Contact contact)
        {
            if (IsPhone(contact))
            {
                List<string> strList = new List<string>();
                string tmp = GetPhone(contact);
                strList.Add($"тел. {tmp} уже есть!");
                strList.Add("Входяший контакт:");
                strList.Add(contact.ToString());
                strList.Add("Контакт в телефонной книге:");
                strList.Add(SearchByTelephone(tmp)!.ToString());
                ErrorConsole.Message(strList);
                return false;
            }
            else
            {
                contactsPhones.Telephones.AddRange(contact.Phone.Telephones);
                return base.Add(contact);
            }
        }

        public override bool Remove(Contact contact)
        {

            if (base.Remove(contact))
            {
                foreach (string item in contact.Phone.Telephones)
                {

                    if (!contactsPhones.RemovePhone(item))
                    {
                        return false;
                    }
                }
                return true;
            }
            return false;
        }
    }
}